frontend/src/metabase/visualizations/components/ChartSettings/ChartSettings.tsx

```diff
@@ -0,0 +1,409 @@
+import { assocIn } from "icepick";
+import { useCallback, useMemo, useState } from "react";
+import { t } from "ttag";
+import _ from "underscore";
+
+import Radio from "metabase/core/components/Radio";
+import CS from "metabase/css/core/index.css";
+import {
+  extractRemappings,
+  getVisualizationTransformed,
+} from "metabase/visualizations";
+import { ChartSettingsFooter } from "metabase/visualizations/components/ChartSettings/ChartSettingsFooter";
+import Visualization from "metabase/visualizations/components/Visualization";
+import { updateSeriesColor } from "metabase/visualizations/lib/series";
+import {
+  getClickBehaviorSettings,
+  getComputedSettings,
+  getSettingsWidgets,
+  updateSettings,
+} from "metabase/visualizations/lib/settings";
+import { getSettingDefinitionsForColumn } from "metabase/visualizations/lib/settings/column";
+import { keyForSingleSeries } from "metabase/visualizations/lib/settings/series";
+import { getSettingsWidgetsForSeries } from "metabase/visualizations/lib/settings/visualization";
+import type Question from "metabase-lib/v1/Question";
+import { getColumnKey } from "metabase-lib/v1/queries/utils/column-key";
+import type { DatasetColumn, VisualizationSettings } from "metabase-types/api";
+
+import ChartSettingsWidgetList from "../ChartSettingsWidgetList";
+import ChartSettingsWidgetPopover from "../ChartSettingsWidgetPopover";
+
+import {
+  ChartSettingsListContainer,
+  ChartSettingsMenu,
+  ChartSettingsPreview,
+  ChartSettingsRoot,
+  ChartSettingsVisualizationContainer,
+  SectionContainer,
+  SectionWarnings,
+} from "./ChartSettings.styled";
+import type {
+  ChartSettingsProps,
+  ChartSettingsWithStateProps,
+  Widget,
+} from "./types";
+
+// section names are localized
+const DEFAULT_TAB_PRIORITY = [t`Data`];
+
+export const ChartSettings = ({
+  initial,
+  settings: propSettings,
+  series,
+  computedSettings: propComputedSettings,
+  onChange,
+  isDashboard = false,
+  noPreview = false,
+  dashboard,
+  dashcard,
+  onDone,
+  onClose,
+  question,
+  className,
+  widgets: propWidgets,
+}: ChartSettingsProps) => {
+  const [currentSection, setCurrentSection] = useState<string | null>(
+    initial?.section || null,
```

Comment:
nit: we could use here and below `??` instead of `||`

Reply:
yes!